examples/ReAct/hotpotqa/inference.py

```diff
@@ -0,0 +1,169 @@
+import wikienv, wrappers
+from reasoners import Reasoner, SearchConfig, WorldModel, Tool
+from reasoners.algorithm import GreedySearch, GreedySearchResult
+from reasoners.lm import Llama3Model
+from reasoners.benchmark import Hotpotqaevaluator
+from reasoners.tools import wikisearch,wikilookup,wikifinish
+import utils
+import copy
+import json
+import fire
+from typing import NamedTuple
+from langchain.tools import StructuredTool
+
+HotpotqaAction = str
+HotpotqaExample = str
+
+class HotpotqaState(NamedTuple):
+    "The state of the Hotpotqa."
+    step_idx: int
+    last_state: str
+    current_state: str
+    action: str
+
+class HotpotqaSearchConfig(SearchConfig):
+
+    def __init__(self,
+                 base_model,
+                 temperature=0):
+        super().__init__()
+        self.base_model = base_model
+        self.temperature = temperature
+
+    def get_actions(self, state):
+        inputs = state.current_state
+        try:
+            outputs = self.base_model.generate([inputs],
+                                                hide_input=True,
+                                                do_sample=True,
+                                                max_new_tokens=512,
+                                                temperature=self.temperature,
+                                                eos_token_id=["Observ","Question"]).text[0]
+        except:
+            return ["exceed maxlength"]
+
+        return [outputs]
+
+    def reward(self, state, action, ans_finished):
+        return 0
+    
+class HotpotqaWorldModel(WorldModel[HotpotqaState, HotpotqaAction, HotpotqaExample]):
+    def __init__(self,
+                 base_model,
+                 toolset,
+                 max_steps: int = 8) -> None:
+        self.base_model = base_model
+        self.toolset = toolset
+        self.max_steps = max_steps
+        self.terminal = False
+
+    def init_state(self):
+        """Initialize the world model.
+        :return: the initial state
+        """
+        current_state = self.prompt["prefix"]+", and Action can be three types: \n"
+        for idx,tool in enumerate(self.toolset):
+            current_state += f"({str(idx+1)}) " + tool.description + "\n"
+        current_state += "Here are some examples.\n" + "".join(self.prompt["examples"])
+        current_state += "Question: "+self.example+'\n'+"Thought 1: "
+        
+        return HotpotqaState(step_idx=1, last_state="", current_state=current_state, action="")
+
+    def step(self, state, action):
+        "Take a step in the world model."
+
+        state = copy.deepcopy(state)
+        current_state = state.current_state
+        step_idx = state.step_idx
+        try:
+            thought, action = action.strip().split(f"\nAction {step_idx}: ")
+            if "Search" in action:
+                new_state = current_state + self.toolset[0](env=self.base_model, step_idx=step_idx, action=action, thought=thought)
```

Comment:
Do not use 0, 1, 2, not to say it's passed to WorldModel as parameter. Define a dataclass or namedtuple